File: src/ai/Embedding.DotNet/OllamaEmbeddingService.cs
using System.Net.Http.Json;
using Microsoft.Extensions.Options;

namespace Embedding.DotNet;

public sealed class OllamaEmbeddingService : IEmbeddingService
{
    private readonly HttpClient _httpClient;
    private readonly EmbeddingOptions _options;

    public OllamaEmbeddingService(HttpClient httpClient, IOptions<EmbeddingOptions> options)
    {
        _httpClient = httpClient ?? throw new ArgumentNullException(nameof(httpClient));
        _options = options?.Value ?? throw new ArgumentNullException(nameof(options));
    }

    public async Task<EmbeddingResponse> GenerateEmbeddingAsync(string text, CancellationToken cancellationToken = default)
    {
        if (string.IsNullOrWhiteSpace(text))
            throw new ArgumentException("Text is required to generate an embedding", nameof(text));

        var model = string.IsNullOrWhiteSpace(_options.Model) ? "nomic-embed-text" : _options.Model;
        using var response = await _httpClient.PostAsJsonAsync("api/embeddings", new OllamaEmbeddingRequest(model, text), cancellationToken);
        response.EnsureSuccessStatusCode();

        var payload = await response.Content.ReadFromJsonAsync<OllamaEmbeddingResponse>(cancellationToken);
        if (payload?.Embedding is null || payload.Embedding.Count == 0)
            throw new InvalidOperationException("Ollama did not return an embedding vector.");

        return new EmbeddingResponse(model, payload.Embedding.Count, payload.Embedding);
    }

    private sealed record OllamaEmbeddingRequest(string Model, string Prompt);

    private sealed record OllamaEmbeddingResponse(IReadOnlyList<float> Embedding);
}

File: src/ai/Embedding.NestJS/ollama-embedding.service.ts
import { Inject, Injectable } from '@nestjs/common';
import { EMBEDDING_OPTIONS, EmbeddingOptions } from './embedding-options';
import { EmbeddingResponse, IEmbeddingService } from './embedding.service.interface';

@Injectable()
export class OllamaEmbeddingService implements IEmbeddingService {
  constructor(
    @Inject(EMBEDDING_OPTIONS)
    private readonly options: EmbeddingOptions
  ) {}

  async generateEmbedding(text: string): Promise<EmbeddingResponse> {
    if (!text || !text.trim()) {
      throw new Error('Text is required to generate an embedding');
    }

    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), this.options.timeoutMs);

    try {
      const response = await fetch(new URL('/api/embeddings', this.options.ollamaBaseUrl), {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({ model: this.options.model, prompt: text }),
        signal: controller.signal,
      });

      if (!response.ok) {
        throw new Error(`Ollama embedding request failed with status ${response.status}`);
      }

      const data = await response.json();
      if (!Array.isArray(data?.embedding) || data.embedding.length === 0) {
        throw new Error('Ollama did not return an embedding vector');
      }

      const embedding = data.embedding.map((value: number) => Number(value));
      return {
        model: this.options.model,
        dimensions: embedding.length,
        embedding,
      };
    } finally {
      clearTimeout(timeout);
    }
  }
}